sql: prohibit phase 2 tenants from setting cluster settings

asubiotto opened the pull request:
The first commit is in #49547 

I wanted to get this PR out for early feedback. Before merging, I want to get #49547 in and then re-run the logic tests to see if any now pass without the 3node-tenant blacklist (at least one should, but it might fail for other reasons).

This commit also adds a testing knob that allows tenants to set in-memory
settings in tests using an updater, which is used in logic tests.

Release note: None (no user-visible change)

Closes #47917

The question of whitelisting settings is not addressed in this PR, @tbg how strongly do you feel about that? AFAIU, whitelisting settings would only be applicable to tests so I'm not sure how useful it is, but haven't thought much about it.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49550)
<!-- Reviewable:end -->

tbg reviewed (approved):
Looks good. I'm fine with not whitelisting settings yet since we're not even exposing any of it outside tests.

> Reviewed 44 of 44 files at r1, 11 of 11 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49550#-:-M8FqqX02DgFs7JEI0a6:b-bllsmu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @nvanbenschoten)

___
*[pkg/server/server_sql_test.go, line 67 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49550#-M8FpJq80lGTsFrhxvZG:-M8FpJq80lGTsFrhxvZH:bmh8fyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cfc99b2e8e3c2dc458008f00593b48cac7f069b/pkg/server/server_sql_test.go#L67)):*
> ```Go
> 	_, err := db.Exec(`SET CLUSTER SETTING sql.defaults.vectorize=off`)
> 	pqErr, ok := err.(*pq.Error)
> 	require.True(t, ok, "expected err to be a *pq.Error but got %v", err)
> ```

nit: if you wanted to log the type, `%T`. Or you could use

`require.IsType(t, (*pg.Error)(nil), err)`
___
*[pkg/sql/logictest/testdata/logic_test/tenant_unsupported, line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49550#-M8Fq63WB2gIT3zsTbHe:-M8Fq63WB2gIT3zsTbHf:b-bhuxlo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cfc99b2e8e3c2dc458008f00593b48cac7f069b/pkg/sql/logictest/testdata/logic_test/tenant_unsupported#L73)):*
> ```
> statement error operation is unsupported
> SELECT * FROM crdb_internal.kv_node_status
> ```

nit: stray edit?


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
Updated logictest logic to set cluster settings on the tenant as well before running the tests and removed the `3node-tenant` blacklist from two tests. I think this is ready to go on after #49547

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49550#-:-M8K_m5l4VSYZzmn6GsP:btg5zws)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/server/server_sql_test.go, line 67 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49550#-M8FpJq80lGTsFrhxvZG:-M8K_PAx2PnX5uGgYsrR:b-e136fz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cfc99b2e8e3c2dc458008f00593b48cac7f069b/pkg/server/server_sql_test.go#L67)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nit: if you wanted to log the type, `%T`. Or you could use

`require.IsType(t, (*pg.Error)(nil), err)`
</blockquote></details>

Added a type verb


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 15 of 15 files at r3, 12 of 12 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49550#-:-M8KiC06Arjd4Zavt0Ax:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @nvanbenschoten)

___
*[pkg/sql/opt/exec/execbuilder/testdata/geospatial, line 3 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49550#-M8KhkWQEoGMm3lYR0fw:-M8KhkWQEoGMm3lYR0fx:bimf1jm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e8a4dace07a38ddfb7e31f78cd4a4d3e65c31db/pkg/sql/opt/exec/execbuilder/testdata/geospatial#L3)):*
> ```
> # This logic test fails in the 3node-tenant configuration because the keys are
> # prefixed with the tenant ID if run by a tenant:
> # https://github.com/cockroachdb/cockroach/issues/49582
> ```

Interesting. @nvanbenschoten any idea how this should work? Make `kvtrace` replace the tenant suffix?


<!-- Sent from Reviewable.io -->

Commit:
sql: prohibit phase 2 tenants from setting cluster settings

This commit also adds a testing knob that allows tenants to set in-memory
settings in tests using an updater, which is used in logic tests.

Release note: None (no user-visible change)

asubiotto commented:
CI failure looks like #47024 

bors r=tbg
